File: Arrays/merge_sorted_array.java
//You are given two integer arrays nums1 and nums2, sorted in non-decreasing order, and two integers
// m and n, representing the number of elements in nums1 and nums2 respectively.

//Merge nums1 and nums2 into a single array sorted in non-decreasing order.

//The final sorted array should not be returned by the function, but instead be stored inside the array nums1.
// To accommodate this, nums1 has a length of m + n, where the first m elements denote the elements that
// should be merged, and the last n elements are set to 0 and should be ignored. nums2 has a length of n.


import java.util.Arrays;

class merge_sorted_array{

    public void merge(int[] nums1, int m, int[] nums2, int n){
        int[] merged=new int[m+n]; //new merged array
        //copying the first m elements from nums1 to the merged array
        if (m >= 0) System.arraycopy(nums1, 0, merged, 0, m);
        //copying the first n elements from nums2 to the merged array
        if (n >= 0) System.arraycopy(nums2, 0, merged, m + 0, n);
        //sorting the merged array
        Arrays.sort(merged);
        //copying the merged array to the nums1 array
        System.arraycopy(merged, 0, nums1, 0, nums1.length);
    }

    public static void main(String[] args){
        int[] nums1 = {1, 2, 3, 0, 0, 0};
        int m = 3, n = 3;
        int[] nums2 ={2, 5, 6};

        merge_sorted_array is = new merge_sorted_array();
        is.merge(nums1, m, nums2, n);
        for (int j : nums1) {
            System.out.print(j + " ");
        }
    }
}

File: Arrays/Merge_Sorted_array.go
Example:
Input: nums1 = [1,2,3,0,0,0], m = 3, nums2 = [2,5,6], n = 3
Output: [1,2,2,3,5,6]

Here's an implementation in Go to merge two sorted arrays:

func merge(nums1 []int, m int, nums2 []int, n int) {
    i := m - 1 // Index of last element in nums1
    j := n - 1 // Index of last element in nums2
    
    // Merge from the end of the arrays to avoid overwriting
    for k := m + n - 1; k >= 0; k-- {
        if j < 0 {
            // No more elements in nums2, nums1 already sorted
            return
        }
        if i >= 0 && nums1[i] > nums2[j] {
            nums1[k] = nums1[i]
            i--
        } else {
            nums1[k] = nums2[j]
            j--
        }
    }
}

The merge function takes in two sorted arrays nums1 and nums2, where nums1 has m elements and nums2 has n elements.
The function then merges nums2 into nums1 in-place, resulting in a single sorted array.

Starting from the end of both arrays, the function compares the last elements of nums1 and nums2, 
and puts the larger one at the end of nums1. This process continues until all elements in nums2 have been merged into nums1.

Note that since nums1 has enough space to accommodate all elements, we don't need to create a new array to store the merged elements.
We can just overwrite the 0 elements at the end of nums1 with the merged elements.
